orc8r/cloud/go/services/directoryd/servicers/servicer.go

```diff
@@ -80,3 +84,74 @@ func (d *directoryLookupServicer) MapSessionIDsToIMSIs(ctx context.Context, req
 
 	return &protos.Void{}, err
 }
+
+// MapHWIDToDirectoryRecordIDs maps {hwid -> directory record IDs}. If existing
+// record IDs already exist for a given hwid, new record IDs will be appended
+// to the existing state.
+func (d *directoryLookupServicer) MapHWIDToDirectoryRecordIDs(ctx context.Context, req *protos.MapHWIDToDirectoryRecordIDsRequest) (*protos.Void, error) {
+	err := req.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to validate request")
+	}
+	for hwid, newRecords := range req.HwidToRecordIDs {
+		existingRecords, err := d.store.GetDirectoryRecordIDsForHWID(req.NetworkID, hwid)
+		// if this is the first record reported for a given hwid then we will
+		// receive an error for non-existent state.
+		if err != nil && err == merrors.ErrNotFound {
+			continue
+		}
+		if err != nil {
+			return &protos.Void{}, errors.Wrap(err, "failed to lookup existing directory record IDs")
+		}
+		allRecordIds := appendIfMissing(existingRecords.Ids, newRecords.Ids)
+		req.HwidToRecordIDs[hwid].Ids = allRecordIds
+	}
+	err = d.store.MapHWIDToDirectoryRecordIDs(req.NetworkID, req.HwidToRecordIDs)
+
+	return &protos.Void{}, err
+}
+
+// GetHWIDToDirectoryRecordIDs returns the directory record IDs mapped to a
+// given hwid. The returned state is pruned of all stale directory records.
+func (d *directoryLookupServicer) GetDirectoryRecordIDsForHWID(ctx context.Context, req *protos.GetDirectoryRecordIDsForHWIDRequest) (*protos.GetDirectoryRecordIDsForHWIDResponse, error) {
+	err := req.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to validate request")
+	}
+	directoryRecords, err := d.store.GetDirectoryRecordIDsForHWID(req.NetworkID, req.Hwid)
+	if err != nil {
+		return nil, err
+	}
+	prunedIDs := []string{}
+	// since state indexing does not handle deletion of stale state, prune
+	// non-existent directory records before returning
+	for _, recordID := range directoryRecords.Ids {
+		_, err := state.GetState(req.NetworkID, orc8r.DirectoryRecordType, recordID, serdes.State)
+		if err != nil && err == merrors.ErrNotFound {
+			continue
+		}
+		if err != nil {
+			return &protos.GetDirectoryRecordIDsForHWIDResponse{}, errors.Wrap(err, "failed to verify if record exists")
+		}
+		prunedIDs = append(prunedIDs, recordID)
+	}
+	res := &protos.GetDirectoryRecordIDsForHWIDResponse{Ids: &protos.DirectoryRecordIDs{Ids: prunedIDs}}
+
+	return res, err
+}
+
+func appendIfMissing(existingIDs []string, newIDs []string) []string {
```

Comment:
Unsure: maybe it makes sense to just return `union (as a set) -> array -> sort` ?

Would be a more stable output, and this helper fn's impl would be simpler